source/framework/analysis/src/TRestPulseShapeAnalysis.cxx

```diff
@@ -0,0 +1,667 @@
+                                                                       : amplitude / averagingPoints;
+        smoothed[i] = sumAvg;
+    }
+
+    // Points at the end, where we can calculate a moving average
+    for (int i = pulseDepth - averagingPoints / 2; i < pulseDepth; i++) smoothed[i] = sumAvg;
+
+    return smoothed;
+}
+
+template std::vector<Float_t> TRestPulseShapeAnalysis::GetSignalSmoothed_ExcludeOutliers<Short_t>(
+    const std::vector<Short_t>& signal, int averagingPoints, Double_t& baseLine, Double_t& baseLineSigma);
+template std::vector<Float_t> TRestPulseShapeAnalysis::GetSignalSmoothed_ExcludeOutliers<Float_t>(
+    const std::vector<Float_t>& signal, int averagingPoints, Double_t& baseLine, Double_t& baseLineSigma);
+
+///////////////////////////////////////////////
+/// \brief Return derivative of a vector of data
+/// points
+///
+template <typename T>
+std::vector<Float_t> TRestPulseShapeAnalysis::GetDerivative(const std::vector<T>& signal) {
+    std::vector<Float_t> derivative(0, signal.size() - 1);
+
+    for (size_t i = 0; i < signal.size() - 1; i++) {
+        derivative[i] = signal[i + 1] - signal[i];
+    }
+
+    return derivative;
+}
+
+template std::vector<Float_t> TRestPulseShapeAnalysis::GetDerivative(const std::vector<Float_t>& signal);
+template std::vector<Float_t> TRestPulseShapeAnalysis::GetDerivative(const std::vector<Short_t>& signal);
+
+///////////////////////////////////////////////
+/// \brief It returns a vector of the data points
+/// that are found over threshold.
+/// The parameters provided to this method are
+/// used to identify those points.
+///
+/// \param thrPar A TVector2 defining two parameters: *pointThreshold* and
+/// *signalThreshold*. Both numbers
+/// define the number of sigmas over the baseline fluctuation, stored in
+/// baseLineSigma. The first parameter,
+/// *pointThreshold*, serves to identify if a single point is over threshold by
+/// satisfying the condition that
+/// is above the baseline by the number of sigmas given in *pointThreshold*.
+/// Once a certain number of
+/// consecutive points have been identified, the parameter *signalThreshold*
+/// will serve to reject the signals
+/// (consecutive points over threshold) that their standard deviation is lower
+/// that *signalThreshold* times
+/// the baseline fluctuation.
+///
+/// \param nPointsOver Only data points with at least *nPointsOver* consecutive
+/// points will be considered.
+///
+/// \param nPointsFlat It will serve to terminate the points over threshold
+/// identification in signals where
+/// we find an overshoot, being the baseline not returning to zero (or its
+/// original value) at the signal tail.
+///
+/// \param baseLine value of the signal baseline calculated beforehand
+///
+/// \param baseLineSigma value of the baseline fluctuation calculated beforehand
+///
+
+template <typename T>
+std::vector<std::pair<Float_t, Float_t> > TRestPulseShapeAnalysis::GetPointsOverThreshold(
```

Comment:
Could return `std::vector<std::pair<Int_t, T> >`?

Reply:
I guess this is related to https://github.com/rest-for-physics/rawlib/pull/100#discussion_r1203847342 
I will try to implement a more generic method.